src/ctypes-foreign-base/ctypes_ffi.ml

```diff
@@ -221,13 +221,78 @@ struct
     fun f ->
       let boxed = cs (Ctypes_weak_ref.make f) in
       let id = Closure_properties.record (Obj.repr f) (Obj.repr boxed) in
-      let funptr = Ctypes_ffi_stubs.make_function_pointer cs' id in
+      Ctypes_ffi_stubs.make_function_pointer cs' id
+
+  let pointer_of_function ~abi ~acquire_runtime_lock ~thread_registration fn =
+    let make_funptr = pointer_of_function_internal ~abi ~acquire_runtime_lock ~thread_registration fn in
+    fun f ->
+      let funptr = make_funptr f in
       (* TODO: use a more intelligent strategy for keeping function pointers
          associated with top-level functions alive (e.g. cache function
          pointer creation by (function, type), or possibly even just by
          function, since the C arity and types must be the same in each case.)
-         See the note by [kept_alive_indefinitely].  *)
+         See the note by [kept_alive_indefinitely].
+
+         [dynamic_funptr_of_fun] below allows for explicit life cycle management
+         and is probably safer to use. *)
       let () = keep_alive funptr ~while_live:f in
       funptr_of_rawptr fn
         (Ctypes_ffi_stubs.raw_address_of_function_pointer funptr)
+
+  type 'a funptr =
+    { mutable gc_root : unit Ctypes.ptr
+    ; fn : 'a Ctypes.static_funptr
+    ; debug_info : string option
+    }
+
+  let free_funptr t =
+    if Ctypes.is_null t.gc_root then
+      (match t.debug_info with
+       | None -> failwith "This funptr was previously freed" | Some debug_info ->
+        failwith (Printf.sprintf "This funptr was previously freed (originally from: %s)"
+                  debug_info))
+    else (
+      Ctypes.Root.release t.gc_root;
+      t.gc_root <- Ctypes.null;
+    )
+
+  let retain_to_avoid_segfaults_when_not_freed_correctly : unit Ctypes.ptr list ref = ref []
+
+  let create_funptr ?debug_info gc_root fn =
+    let t = { debug_info; gc_root = Ctypes.Root.create gc_root; fn } in
+    Gc.finalise (fun t ->
+      if Ctypes.is_null t.gc_root then
+        ()
+      else (
+        (match t.debug_info with
+         | None -> ()
+         | Some debug_info ->
+          Printf.eprintf
+            "WARN: a ctypes function pointer was not explicitly released.\n\
+             (originally allocated in: %s)\n\
+             Releasing a function pointer or the associated ocaml closure while \n\
+             the function pointer is still in use will cause segmentation faults.\n\
+             Please call [Foreign.free_funptr] explicitly when it is no longer needed.\n\
+             To avoid a segmentation fault we are preventing this function pointer from\n\
+             being garbage collected. Please use [Foreign.free_funptr].\n%!"
```

Comment:
I am concerned about the printing to stderr; this will just get lost in the noise for many uses of Ctypes.

Why not follow the pattern of raising an exception (used in the case of double free) and raise one from the finaliser?  This will almost certainly terminate the program due to it interrupting the mutator, but cleanly and without heap corruption.

Alternatively, the library could also register a global exception hook to be called with the value, permitting the Printf to be overridden (e.g. by the `Logs` library, or some other system-specific action).

Reply:
Thanks, fixed to use a callback for reporting the error.
By default this will still print to stderr, but you can change the behaviour, to e.g. use `Logs`
```ocaml
Foreign.report_leaked_funptr := (fun msg -> Logs.error (fun m -> m "%s" msg))
```
(I've never used `Logs` so the above probably isn't quite right)